### DISCUSSION TITLE
How to configure to send task logs to datadog?

### DISCUSSION THREAD

**Discussion by seaguest** [apache/airflow#30956] [General]
Hello,

I deployed airflow with helm in Kubernetes
```
helm upgrade --install datadog-monitoring \
    --set datadog.apiKey="xxx" \
    --set clusterAgent.enabled=true \
    --set clusterAgent.token="xxx" \
    --set datadog.criSocketPath=null \
    --set datadog.logs.enabled=true \
    --set datadog.logs.containerCollectAll=true \
    --set datadog.containerExcludeLogs="name:dashboard-metrics-scraper name:kubernetes-dashboard" \
    --set datadog.dogstatsd.port=8125 \
    --set datadog.dogstatsd.useHostPort=true \
    --set datadog.dogstatsd.nonLocalTraffic=true \
    --set datadog.apm.portEnabled=true \
    --set datadog.apm.port=8126 \
    datadog/datadog
```

All other configuration files are default ones.

I can see some logs in datadog
```
[2023-04-29 11:51:26,584: INFO/ForkPoolWorker-1] Task airflow.executors.celery_executor.execute_command[2a8bbfc1-4c3c-4123-bdc8-36dc6fe7799d] succeeded in 59.32762966118753s: None
```

However the task logs is not sent to datadog, I can see `/opt/airflow/logs` does have a lot of task logs but they are never sent to datadog, what is the right way to do that?

It seems only stdout logs are uploaded to datadog, what is missing here to upload task logs to datadog?

**Comment by potiuk** [comment]
https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#statsd-datadog-enabled ?

**Reply by seaguest** [reply]
Here is the config for metric, I think we do have metric uploaded to datadog, only the task logs not uploaded.
Not sure what I have missed, should I try to write logs directly to STDOUT like https://stackoverflow.com/questions/65651301/how-to-output-airflows-scheduler-log-to-stdout-or-s3-gcs?
```
		[metrics]
		statsd_host = airflow-statsd
		statsd_on = True
		statsd_port = 9125
		statsd_prefix = airflow
```

**Reply by potiuk** [reply]
Ah. you mean task logs? You need to write your own logger for that if you want. I do not think there is built-in logger for datadog:  https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/logging.html but you can write your own (and possibly contribute it back if you want).

**Reply by seaguest** [reply]
@potiuk 
datadog agent only send logs from STDOUT to datadog, while task logs are only written to local files.
I thought there should be some easy way to forward these local logs to datadog too, because this is a very common use case.
`Logging for Tasks`  are only for uploading to remote server, what I want is to write to STDOUT.

I'll try the solution in my above stackoverflow link.

**Reply by potiuk** [reply]
Likely you do not understand how tasks work usually (Local Executor/Celery Executor). 

When you run tasks airflow forks them and it spawns multiple sub-processes that are running multiple tasks (all spawned from single process which has single stdout). If you **JUST** forward stdout from that process (even if you redirect the task output to stdout) you will have undistinguishable interleaved log lines from multiple tasks. Your only option is to use task logging handler from python and you need task logging handler to know which task each log comes from.  This is what task handlers do.

**Reply by Sarthakgarg17** [reply]
@seaguest  i am also facing this same issue. did you find its solution?

**Reply by andrey-ch-dev** [reply]
from https://docs.datadoghq.com/data_jobs/airflow/?tab=kubernetes#setup but it requires https://docs.datadoghq.com/containers/kubernetes/log/?tab=helm#from-a-container-local-log-file as well

```
ad.datadoghq.com/base.logs: '[{"type": "file", "path": "PATH_TO_AIRFLOW_LOGS/*/*/*/*.log", "source": "airflow"}]'
```
